<a href="https://colab.research.google.com/github/Rastaban19/Temporary/blob/main/InformaAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#DataFrame

In [1]:
import pandas as pd

In [2]:
from google.colab import drive

In [4]:
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
path = "/content/exhibitors_informa.csv"
df = pd.read_csv(path)

FileNotFoundError: [Errno 2] No such file or directory: '/content/exhibitors_informa.csv'

In [ ]:
df

#EDA

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1243 entries, 0 to 1242
Data columns (total 35 columns):
 #   Column                                     Non-Null Count  Dtype 
---  ------                                     --------------  ----- 
 0   Name                                       1243 non-null   object
 1   Description                                1207 non-null   object
 2   Country                                    1243 non-null   object
 3   Email                                      1243 non-null   object
 4   Phone                                      1243 non-null   object
 5   Mobile                                     1243 non-null   object
 6   Phone Number                               1243 non-null   object
 7   Address                                    1243 non-null   object
 8   Logos                                      661 non-null    object
 9   Products                                   600 non-null    object
 10  Documents                           

In [ ]:
df.shape

(1243, 35)

In [ ]:
df.describe()

Name           Description Country           Email  \
count              1243                  1207    1243            1243   
unique             1242                   748      40             667   
top     Informa Markets  No description found   China  No email found   
freq                  2                   460     458             577   

                 Phone           Mobile    Phone Number           Address  \
count             1243             1243            1243              1243   
unique             514              503              23               296   
top     No phone found  No mobile found  No phone found  No address found   
freq               730              741            1221               943   

                                                    Logos  \
count                                                 661   
unique                                                661   
top     https://img.swapcard.com/?u=https%3A%2F%2Fstat...   
freq                                                    1   

                                                 Products  ...  \
count                                                 600  ...   
unique                                                 11  ...   
top     [object Object],[object Object],[object Object...  ...   
freq                                                  292  ...   

       Healthcare Infrastructure & Assets IT Systems / Solutions  \
count                                1243                   1243   
unique                                  1                      1   
top                             Not found              Not found   
freq                                 1243                   1243   

       IT Systems, Solutions & Digital Health Imaging & Diagnostics  \
count                                    1243                  1243   
unique                                     57                     1   
top                                 Not found             Not found   
freq                                     1158                  1243   

       Medical Equipment & Devices Orthopedic & Physiotherapy/Rehabilitation  \
count                         1243                                      1243   
unique                           1                                         1   
top                      Not found                                 Not found   
freq                          1243                                      1243   

       Pharma/Nutrition    Imaging  \
count              1243       1243   
unique               19         54   
top           Not found  Not found   
freq               1173       1118   

                                             Social Media           Website  
count                                                 436              1243  
unique                                                436               640  
top     https://www.linkedin.com/company/bonetech-medi...  No website found  
freq                                                    1               583  

[4 rows x 35 columns]

In [ ]:
df.columns.tolist()

['Name',
 'Description',
 'Country',
 'Email',
 'Phone',
 'Mobile',
 'Phone Number',
 'Address',
 'Logos',
 'Products',
 'Documents',
 'Document Links',
 'Product Images',
 'Product Links',
 'Billboard Image',
 'Billboard Video',
 'Nature of Business',
 'Interested to Connect With',
 'Laboratory',
 'Orthopaedic Devices',
 'Physiotherapy/Rehabilitation/Mobility',
 'Country Coverage',
 'Disposables & Consumables',
 'Medical Equipment',
 'Healthcare & General Services',
 'Healthcare Infrastructure & Assets',
 'IT Systems / Solutions',
 'IT Systems, Solutions & Digital Health',
 'Imaging & Diagnostics',
 'Medical Equipment & Devices',
 'Orthopedic & Physiotherapy/Rehabilitation',
 'Pharma/Nutrition',
 'Imaging',
 'Social Media',
 'Website']

In [ ]:
df.nunique()
#This is the number of unique values per each element/observation of the dataframe.
#We recall the df shape (1243, 35)
#Name is closer to the lenght so we can identify it as PK/id of the elements. There is one missing value, which is strange,
#but we might improve the quality of the dataset by removing this row.

Name                                         1242
Description                                   748
Country                                        40
Email                                         667
Phone                                         514
Mobile                                        503
Phone Number                                   23
Address                                       296
Logos                                         661
Products                                       11
Documents                                      70
Document Links                                 71
Product Images                                568
Product Links                                 600
Billboard Image                                36
Billboard Video                                 1
Nature of Business                             22
Interested to Connect With                    270
Laboratory                                    100
Orthopaedic Devices                            28


From here I can either start data reshaping with a copy of the df or I could procede with the identification of the necessar information.

In [ ]:
df.isnull().sum()

Name                                            0
Description                                    36
Country                                         0
Email                                           0
Phone                                           0
Mobile                                          0
Phone Number                                    0
Address                                         0
Logos                                         582
Products                                      643
Documents                                    1172
Document Links                               1172
Product Images                                643
Product Links                                 643
Billboard Image                                 0
Billboard Video                                 0
Nature of Business                              0
Interested to Connect With                      0
Laboratory                                      0
Orthopaedic Devices                             0


!!! My intuition was wrong. We have 1242 insted of 1243 names because there are two companies with the same name! So there are no null values. But this makes it impossibile to use this value as a FK, we can fix in two ways:
1) Assign an unique ID to each observation
2) Investigate why only these two values come with the same name, might be a typo.

In [ ]:
#df.isnull().sum().sum()

In [ ]:
dfcopy = df.copy()

In [ ]:
dfcopy

Name  \
0                      BONETECH MEDISYS PRIVATE LIMITED   
1                                                 Crein   
2                         F.L. Medical Srl Unipersonale   
3                                           FULWELL LLC   
4                                       Gpc Medical Ltd   
...                                                 ...   
1238                     Zibo Sunrise Medical Co., Ltd.   
1239                         Zilla Technology PTE. LTD.   
1240                                      ZOE BIO, Inc.   
1241                                  ZOLL MEDICAL CORP   
1242   Changzhou Intl. Trade & Enterprises Cooperati...   

                                            Description  \
0     Introducing Bonetech Medisys, a premier manufa...   
1     We are the leading regulatory affairs firm in ...   
2     Research, evolution, excellence, quality. The ...   
3     Baisheng Medical Co.,Ltd, OBS set its producti...   
4     Introducing GPC Medical, a reputable global le...   
...                                                 ...   
1238  SUNRISE MEDICAL is a global leader in design a...   
1239                               No description found   
1240                               No description found   
1241  ZOLL develops and markets medical devices and ...   
1242  Changzhou Intl. Trade & Enterprises Cooperativ...   

                       Country                        Email  \
0                        India  gaurang@bonetechmedisys.com   
1                      Ecuador            info@crein.com.ec   
2                        Italy           info@flmedical.com   
3     United States of America             sales@bs0750.com   
4                        India          info@gpcmedical.com   
...                        ...                          ...   
1238                     China           info@sunrisemed.cn   
1239                 Singapore               No email found   
1240               South Korea               No email found   
1241  United States of America     LatamSales.zoll@zoll.com   
1242                     China        citec@citec-group.com   

                   Phone             Mobile    Phone Number  \
0         0091 957407704      +91 957407704  No phone found   
1      00593 98 041 6519    +593 2 222 8813  No phone found   
2         No phone found    No mobile found  No phone found   
3     0086 137 1726 4770   +86 750 662 8113  No phone found   
4      0091 11 4322 2600   +91 11 4322 2600  No phone found   
...                  ...                ...             ...   
1238  0086 155 5332 5069   +86 533 216 8162  No phone found   
1239      No phone found    No mobile found  No phone found   
1240      No phone found    No mobile found  No phone found   
1241      No phone found    No mobile found  No phone found   
1242  0086 519 8680 3321  +86 519 8680 3321  No phone found   

                                                Address  \
0     Survey No.16 , Bibipura Road, Geratnagar, Near...   
1     Oficina Principal29, Leonidas Plaza Gutierrez,...   
2                                      No address found   
3                                      No address found   
4                                          See all (17)   
...                                                 ...   
1238                                   No address found   
1239                                   No address found   
1240                                   No address found   
1241                                   No address found   
1242                                   No address found   

                                                  Logos  \
0                                                   NaN   
1     https://img.swapcard.com/?u=https%3A%2F%2Fstat...   
2     https://img.swapcard.com/?u=https%3A%2F%2Fcdn-...   
3     https://img.swapcard.com/?u=https%3A%2F%2Fcdn-...   
4                                                   NaN   
...                                                 ... 

##sqlite

In [ ]:
import sqlite3

In [ ]:
def pd_to_sqlDB(input_df: dfcopy,
                table_name: str,
                db_name: str = 'default.db') -> None:

    '''Take a Pandas dataframe `input_df` and upload it to `table_name` SQLITE table
    Args:
        input_df (pd.DataFrame): Dataframe containing data to upload to SQLITE
        table_name (str): Name of the SQLITE table to upload to
        db_name (str, optional): Name of the SQLITE Database in which the table is created.
                                 Defaults to 'default.db'.
    '''

In [ ]:
    # Step 1: Setup local logging
    import logging
    logging.basicConfig(level=logging.INFO,
                        format='%(asctime)s %(levelname)s: %(message)s',
                        datefmt='%Y-%m-%d %H:%M:%S')

In [ ]:
    # Step 2: Find columns in the dataframe
    cols = dfcopy.columns
    cols_string = ','.join(cols)
    val_wildcard_string = ','.join(['?'] * len(cols))

In [ ]:
    # Step 3: Connect to a DB file if it exists, else crete a new file
    con = sqlite3.connect(db_name)
    cur = con.cursor()
    logging.info(f'SQL DB {db_name} created')

NameError: name 'db_name' is not defined

In [ ]:
    # Step 4: Create Table
    sql_string = f"""CREATE TABLE {table_name} ({cols_string});"""
    cur.execute(sql_string)
    logging.info(f'SQL Table {table_name} created with {len(cols)} columns')

    # Step 5: Upload the dataframe
    rows_to_upload = input_df.to_dict(orient='split')['data']
    sql_string = f"""INSERT INTO {table_name} ({cols_string}) VALUES ({val_wildcard_string});"""
    cur.executemany(sql_string, rows_to_upload)
    logging.info(f'{len(rows_to_upload)} rows uploaded to {table_name}')

    # Step 6: Commit the changes and close the connection
    con.commit()
    con.close()

#Missing Values

###Missing Countries


In [ ]:
country_test = dfcopy['Country'].unique()
country_test.sort()
country_test.tolist()

In [ ]:
dfcopy['Country'].value_counts()["No country found"]  # Count occurrences of "No country found" in the "Country" column

In [ ]:
dfcopy['Country'].value_counts()

In [ ]:
MissingCountry = dfcopy[dfcopy["Country"].isin(["NULL", "No country found"])]

In [ ]:
MissingCountry

In [ ]:
MissingCountry.drop(columns=MissingCountry.columns.difference(["Name","Country"]),inplace=True)

In [ ]:
MissingCountry

###Missing Phone Numbers

In [ ]:
MissingPhone = dfcopy[dfcopy["Phone"].isin(["NULL", "No phone found"])]

In [ ]:
MissingPhone

In [ ]:
MissingPhone.shape

In [ ]:
MissingPhone.drop(columns=MissingPhone.columns.difference(["Name","Phone"]),inplace=True)

In [ ]:
MissingPhone
#This interactive table can be easily downloaded as a csv.

###Missing Emails

In [ ]:
MissingEmail = dfcopy[dfcopy["Email"].isin(["NULL", "No email found"])]
MissingEmail.drop(columns=MissingEmail.columns.difference(["Name","Email"]),inplace=True)
MissingEmail

###Missing Addresses

In [ ]:
MissingAddress = dfcopy[dfcopy["Address"].isin(["NULL", "No address found"])]
MissingAddress.drop(columns=MissingAddress.columns.difference(["Name","Address"]),inplace=True)
MissingAddress

###Missing Nature of Businesses

In [ ]:
dfcopy["Nature of Business"]

In [ ]:
b_test = dfcopy['Nature of Business'].unique()
b_test.sort()
b_test.tolist()

In [ ]:
Missingb = dfcopy[dfcopy["Nature of Business"].isin(["NULL", "Not found"])]
Missingb.drop(columns=Missingb.columns.difference(["Name","Nature of Business"]),inplace=True)
Missingb

###Missing Websites

In [ ]:
dfcopy["Website"]

In [ ]:
Missingw = dfcopy[dfcopy["Website"].isin(["NULL", "No website found"])]
Missingw.drop(columns=Missingw.columns.difference(["Name","Website"]),inplace=True)
Missingw

In [ ]:
c = MissingCountry.value_counts().sum()
p = MissingPhone.value_counts().sum()
e = MissingEmail.value_counts().sum()
a = MissingAddress.value_counts().sum()
b = Missingb.value_counts().sum()
w = Missingw.value_counts().sum()

In [ ]:
Country = ["Country"]

In [ ]:
C = len(Country.axes[0])
C

In [ ]:
import seaborn as sns